### PR TITLE
Optimize Test Suite

### PR DESCRIPTION
**Pull request by esc**
This implements some additional instrumentation for the test-suite and adds
some additional Python 3 tricks to make the tests go faster.

### CODE REVIEW

**Comment by esc** [comment]
It is important to note here, that we could not use the `chunksize` argument of the `map` function. According to the docs, this should make the tests run much quicker. Unfortunately, the tests occasionally timeout and the only way to detect this is by setting the `chunksize` to `1` as this adds an optional `timeout` parameter to the `next` method of the iterator. See the following for more details:

https://docs.python.org/3.7/library/multiprocessing.html

**Comment by esc** [comment]
Next steps would be to take a closer look at which tests could be 'deserialized' i.e. moved from the serial test-suite to the parallel one.

**Comment by esc** [comment]
I should probably also note that, this will give an increase of about 10% on my 12-core MBP.

File: numba/testing/main.py [hunk 1]
```diff
@@ -639,13 +639,21 @@ def _run_inner(self, result):
...
+                              for i in range(0, len(self._ptests), chunk_size)]
+            target = splitted_tests
+        elif six.PY3:
+            target = [self._ptests]
+        t = time.time()
+        for tests in target:
+            if six.PY2:
+                pool = multiprocessing.Pool(self.nprocs)
+            elif six.PY3:
+                pool = multiprocessing.get_context('spawn').Pool(self.nprocs,
```
Comment (stuartarchibald) [inline numba/testing/main.py]: Did you try forkserver too, was it successful?

File: numba/testing/main.py [hunk 2]
```diff
@@ -692,6 +706,8 @@ def _run_parallel_tests(self, result, pool, child_runner, tests):
 
     def run(self, test):
         self._ptests, self._stests = _split_nonparallel_tests(test)
+        print("Total parallel test functions: '{}'".format(len(self._ptests)))
```
Comment (stuartarchibald) [inline numba/testing/main.py]:
```suggestion
        print("Number of parallel test functions: '{}'".format(len(self._ptests)))
```
(same below)

File: numba/testing/main.py (hunk 1 above)
Comment (esc) [reply numba/testing/main.py]: no, I haven't tried yet.

File: numba/testing/main.py [hunk 3]
```diff
@@ -639,13 +639,21 @@ def _run_inner(self, result):
...
+            splitted_tests = [self._ptests[i:i + chunk_size]
+                              for i in range(0, len(self._ptests), chunk_size)]
+            target = splitted_tests
+        elif six.PY3:
+            target = [self._ptests]
+        t = time.time()
+        for tests in target:
+            if six.PY2:
+                pool = multiprocessing.Pool(self.nprocs)
+            elif six.PY3:
```
Comment (stuartarchibald) [inline numba/testing/main.py]: Perhaps this branching needs to be based on  `< 3.4` as that was when `.get_context` was added?

Comment (esc) [reply numba/testing/main.py]: yes, changed and pushed.

File: numba/testing/main.py (hunk 1 above)
Comment (esc) [reply numba/testing/main.py]: so I tried this, and it works.

**Comment by esc** [comment]
Looks like the `Linux py37_np114`  build is crashing due to these:

```
======================================================================
ERROR: test_stack_effect_error_npm (numba.tests.test_dataflow.TestDataFlow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_dataflow.py", line 175, in test_stack_effect_error_npm
    self.test_stack_effect_error(no_pyobj_flags)
  File "/home/vsts/work/1/s/numba/tests/test_dataflow.py", line 169, in test_stack_effect_error
    cr = compile_isolated(pyfunc, (types.int32,), flags=flags)
  File "/home/vsts/work/1/s/numba/compiler.py", line 157, in compile_isolated
    flags, locals)
  File "/home/vsts/work/1/s/numba/compiler.py", line 940, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/vsts/work/1/s/numba/compiler.py", line 372, in compile_extra
    return self._compile_bytecode()
  File "/home/vsts/work/1/s/numba/compiler.py", line 871, in _compile_bytecode
    return self._compile_core()
  File "/home/vsts/work/1/s/numba/compiler.py", line 858, in _compile_core
    res = pm.run(self.status)
  File "/home/vsts/work/1/s/numba/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/vsts/work/1/s/numba/compiler.py", line 254, in run
    raise patched_exception
  File "/home/vsts/work/1/s/numba/compiler.py", line 245, in run
    stage()
  File "/home/vsts/work/1/s/numba/compiler.py", line 496, in stage_nopython_frontend
    self.locals)
  File "/home/vsts/work/1/s/numba/compiler.py", line 1089, in type_inference_stage
    infer.build_constraint()
  File "/home/vsts/work/1/s/numba/typeinfer.py", line 846, in build_constraint
    self.constrain_statement(inst)
  File "/home/vsts/work/1/s/numba/typeinfer.py", line 1125, in constrain_statement
    self.typeof_assign(inst)
  File "/home/vsts/work/1/s/numba/typeinfer.py", line 1188, in typeof_assign
    self.typeof_global(inst, inst.target, value)
  File "/home/vsts/work/1/s/numba/typeinfer.py", line 1285, in typeof_global
    typ = self.resolve_value_type(inst, gvar.value)
  File "/home/vsts/work/1/s/numba/typeinfer.py", line 1206, in resolve_value_type
    return self.context.resolve_value_type(val)
  File "/home/vsts/work/1/s/numba/typing/context.py", line 358, in resolve_value_type
    ty = typeof(val, Purpose.constant)
  File "/home/vsts/work/1/s/numba/typing/typeof.py", line 30, in typeof
    ty = typeof_impl(val, c)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/functools.py", line 827, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/vsts/work/1/s/numba/typing/typeof.py", line 96, in typeof_type
    from numba.typed import Dict
  File "/home/vsts/work/1/s/numba/typed/__init__.py", line 3, in <module>
    from .typeddict import Dict
  File "/home/vsts/work/1/s/numba/typed/typeddict.py", line 4, in <module>
    from collections import MutableMapping
  File "<frozen importlib._bootstrap>", line 1032, in _handle_fromlist
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/collections/__init__.py", line 52, in __getattr__
    DeprecationWarning, stacklevel=2)
DeprecationWarning: Failed in nopython mode pipeline (step: nopython frontend)
Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
```

**Comment by esc** [comment]
I am unable to reproduce this locally, will wait for #3935 to be merged and then try this one again.

**Comment by esc** [comment]
@stuartarchibald do you have any hints as to how to discover tests that are marked as 'serial' but could be done in 'parallel'?

**Comment by stuartarchibald** [comment]
Typically, if a test mutates something in a way that might impact another process running a different test then the first test needs to be serial, for example, a test writing to an on-disk cache. Since all the threading layer work, it might be work seeing if the tests which use threads (i.e. parallel target tests) can now also run concurrently (they exhibit little to no workload in practice so this should be fine, so long as runtime can handle it).

**Comment by seibert** [comment]
Since we were looking at this bit of code this morning:
```
https://github.com/numba/numba/blob/master/numba/caching.py#L190-L210
```
I wonder if we could have most all of the caching tests use a custom, test-specific temp location and then only have a handful of tests check the other caching locations.

**Comment by esc** [comment]
Merged master, giving this another run through CI now.

**Comment by stuartarchibald** [comment]
I think the failure here https://dev.azure.com/numba/numba/_build/results?buildId=1595 (Linux py37_np116_tbb) may well be genuine.

**Comment by esc** [comment]
@stuartarchibald what does the suffix `tbb` stand for?

**Comment by stuartarchibald** [comment]
@esc Threading Building Blocks https://01.org/tbb, suggested reading: http://numba.pydata.org/numba-doc/latest/user/threading-layer.html

**Comment by esc** [comment]
@stuartarchibald yeah, I concur the tbb failure is legit.

**Comment by esc** [comment]
It was 681 commits behind `master`, let's re-roll this.

**Comment by esc** [comment]
Closing in favor of #4781 and setting to 'abandoned'.

**Pull request closed without merging**